modules/build/src/main/scala/scala/build/CrossSources.scala

```diff
@@ -403,7 +405,39 @@ object CrossSources {
     fromInputs ++ fromSources ++ fromSourcesWithRequirements
   }
 
-  private def resolveInputsFromSources(sources: Seq[Positioned[os.Path]], enableMarkdown: Boolean) =
+  // TODO: reuse existing one? e.g. scala.cli.commands.shared.SharedOptions.coursierCache
+  lazy val fileCache: FileCache[coursier.util.Task] = FileCache()
+
+  private def downloadFile(pUri: Positioned[java.net.URI]) =
+    import scala.build.options.ScalaVersionUtil.fileWithTtl0
+    val artifact = Artifact(pUri.value.toString).withChanging(true)
+    fileCache.fileWithTtl0(artifact)
+      .left
+      .map(err =>
+        new MalformedDirectiveError(err.describe, pUri.positions)
+      ) // TODO: better error type
+      .map(f => os.read.bytes(os.Path(f, Os.pwd))).map(content =>
+        Seq(Virtual(pUri.value.toString, content))
+      )
```

Comment:
This is very similar to `SharedOptions.downloadInputs`... I wonder if the function itself could be passed all the way here, rather than the cache